File: packages/yt/lib/src/activities.dart
import 'package:yt/src/youtube_api_helper.dart';
import 'package:yt/yt.dart';

import 'provider/data/activities.dart';

/// An activity resource contains information about an action that a particular channel, or user, has taken on YouTube.
class Activities extends YouTubeApiHelper {
  final ActivitiesClient _rest;

  /// Returns a list of channel activity events that match the request criteria.
  Activities({required super.dio, super.token, super.apiKey})
    : _rest = ActivitiesClient(dio);

  /// Returns a list of channel activity events that match the request criteria.
  /// For a channel, the API returns a list of that channel's recent activities.
  Future<ActivityListResponse> list({
    String part = 'snippet,contentDetails',
    List<String> partList = const [],
    String? channelId,
    bool? mine,
    int? maxResults,
    String? pageToken,
    String? publishedAfter,
    String? publishedBefore,
    String? regionCode,
  }) async => _rest.list(
    apiKey,
    accept,
    buildParts(partList, part),
    channelId: channelId,
    mine: mine,
    maxResults: maxResults,
    pageToken: pageToken,
    publishedAfter: publishedAfter,
    publishedBefore: publishedBefore,
    regionCode: regionCode,
  );
}

File: packages/yt_cli_js/src/commands/activities.ts
import { Command } from 'commander';
import type { YtCliJsHandle, YtCliJsNamespace } from '../types.js';
import { getRuntime } from '../runtime.js';

export function registerActivitiesCommand(program: Command): void {
  const activities = program
    .command('activities')
    .description('YouTube activity operations');

  activities
    .command('list')
    .description('List YouTube activities')
    .option('--part <parts>', 'Resource parts to include', 'snippet,contentDetails')
    .option('--channel-id <id>', 'Channel ID')
    .option('--mine', 'List activities for the authenticated user')
    .option('--max-results <n>', 'Maximum number of results', '5')
    .option('--page-token <token>', 'Page token for pagination')
    .option('--published-after <date>', 'Filter by publish date (RFC 3339)')
    .option('--published-before <date>', 'Filter by publish date (RFC 3339)')
    .option('--region-code <code>', 'Region code (ISO 3166-1 alpha-2)')
    .action(async (opts) => {
      try {
        const apiKey = program.opts().apiKey || process.env.YT_API_KEY;
        if (!apiKey) {
          console.error('Error: API key required. Use --api-key or set YT_API_KEY.');
          process.exit(1);
        }

        const ns: YtCliJsNamespace = await getRuntime();
        const logLevel = program.opts().logLevel;
        const handle: YtCliJsHandle = await ns.withApiKey(apiKey, logLevel);

        try {
          const result = await handle.activitiesList(
            opts.channelId,
            opts.mine ?? false,
            opts.part,
            parseInt(opts.maxResults, 10),
            opts.pageToken,
            opts.publishedAfter,
            opts.publishedBefore,
            opts.regionCode,
          );
          console.log(JSON.stringify(result, null, 2));
        } finally {
          handle.close();
        }
      } catch (err: any) {
        console.error(`Error: ${err.message ?? err}`);
        process.exit(1);
      }
    });
}
